Add solidus-preview

josiahsprague opened the pull request:
A continuation of https://github.com/blocks/post-preview/pull/1

josiahsprague commented:
- [ ] Add optional `element` parameter to `render()` method to simplify implementation of rendering returned HTML into a page.
- [ ] Development of `SolidusClient.renderWordPressPreview()` on https://github.com/solidusjs/solidus-preview/pull/1 (naming?)

josiahsprague commented:
Thanks @joanniclaborde for the help in making design decisions for this. :) 

This is ready for some more review. Usability is exactly the same as before. You can see it in action here: https://github.com/SparkartGroupInc/jasonaldean.com/pull/77

cc @pushred

joanniclaborde commented on lib/wordpress.js:
```diff
@@ -0,0 +1,48 @@
+var _ = require('underscore');
+var params = require('querystring').parse(window.location.search.split('?').pop());
+
+module.exports = {
```
`module.exports` is already an object, for readability and consistency with the rest of the package, I suggest `module.exports.getPostPreviewView = function (view) {`.

josiahsprague replied:
I did it this way to leave room for other functions here, like `.getProtectedPostView()` on line 44. Rather than repeating `module.exports` at various arbitrary points throughout the file, I think it's cleaner to do it this way.

joanniclaborde commented on lib/wordpress.js:
```diff
@@ -0,0 +1,48 @@
+    if (_.isFunction(view)) {
+      view = {template: view};
+    }
+    if (_.isFunction(view.preprocessor)) {
```
This line will blow up if `view` is undefined.

josiahsprague replied:
Added a check for `view` on line 8.

joanniclaborde replied:
Are you sure a view is required?
```javascript
var view = getPostPreviewView();
view.template = ...;
```

josiahsprague replied:
If there's no `view` ∴ no template, then we don't have all of the arguments we need to return a usable `view` to pass into `render()`. I can't imagine a scenario where we'd call `getPostPreviewView()` without any arguments.

pushred replied:
Wouldn't hurt to throw just in case if someone did though, telling them what's required rather than silently failing.

joanniclaborde commented on lib/wordpress.js:
```diff
@@ -0,0 +1,48 @@
+      },
+      preprocessor: function(context) {
+        var lastRevision = context.resources.revisions[0];
+        context.resources.wordpress = _.extend(context.resources.wordpress, lastRevision);
```
No need for the assignment here, `_.extend` modifies its first argument.

joanniclaborde commented on lib/wordpress.js:
```diff
@@ -0,0 +1,48 @@
+        var lastRevision = context.resources.revisions[0];
+        context.resources.wordpress = _.extend(context.resources.wordpress, lastRevision);
+        if (originalPreprocessor) { context = originalPreprocessor(context); }
+        return context.resources.wordpress;
```
Won't this break any other preprocessor or template that needs the context?

josiahsprague replied:
This preprocessor is only used for the `wordpress.getPostPreviewView()` function, so it will only be used by WordPress post templates. Those typically [just have the WordPress resource passed in as the context](https://github.com/SparkartGroupInc/jasonaldean.com/blob/master/views/news/%7Bcategory%7D/%7Bid%7D/%7Bslug%7D.hbs#L26). This preprocessor will run other preprocessors first.

joanniclaborde commented on lib/wordpress.js:
```diff
@@ -0,0 +1,48 @@
+    };
+
+    // Return view extended with preview object
+    return _.extend(view, preview);
```
Again, this will modify `view`, as a side effect of the function. You can do `return _.extend({}, view, preview);` instead.

josiahsprague replied:
Is modifying `view` as a side effect bad here? It's only within this function's scope, and I think it's more readable, and clear what's we're doing this way.

joanniclaborde replied:
This will modify the caller's object too:
```javascript
function getPostPreviewView(view) {view.test = true; return view}
var view = {};
getPostPreviewView(view);
console.log(view);
> Object {test: true}
```

If the caller really wants to change his variable he should do
```javascript
view = getPostPreviewView(view);
```

joanniclaborde commented:
You forgot the tests!

pushred commented:
My 2 cents on the `replace` method — it would be better named `replaceWith` for consistency with jQuery, these subtle differences in libraries can really mess with your head in dev. I'd rather not add DOM  manipulation features like that at all but the async nature of `render` does make it verbose to work with otherwise. It would be good to round that out with `appendTo` and `prependTo` for the scenarios where those will inevitably be needed to avoid DOM gymnastics.

josiahsprague commented:
I amended some changes to address the feedback above. I'm going to add tests, as long as the patterns I'm using here are looking stable to everyone. :)

I'm not sure that a `replaceWith()` method would make sense here. The jQuery method is called on an element and takes the HTML as an argument, which is essentially the reverse of what's happening here. My rationale behind adding DOM manipulation features was just to simplify the usage of the post preview feature across Solidus sites. In reality, that's a completely separate feature from post previews though, and can probably be addressed through a separate PR/discussion at a later date. For now, it's not **that** complicated to do;
```
solidus_client.render(view, params).end(function(err, html) {
  document.querySelector('.story.module').outerHTML = html;
});
```
The only downside is potential inconsistency with how things are handled when `err` is true.

joanniclaborde commented:
If you do add a `replace` method, it should accept a second callback argument, so the user can check if an error happened:
```javascript
solidus_client.render(view).replace('element', function (err) {...});
```

pushred commented on lib/wordpress.js:
```diff
@@ -0,0 +1,48 @@
+    // Return false if we don't have everything we need for a preview
+    var isPreviewActive = params._wp_json_nonce && params.is_preview;
+    if (!isPreviewActive) {
+      console.info('Preview is not active.');
```
This will throw errors in IE 9 (and below), and it's not standard JS still so I think it's best to leave it out.

pushred commented:
> If you do add a replace method, it should accept a second callback argument, so the user can check if an error happened

This kinda defeats the convenience of this method to me. How about some default error handling with messaging output as a simple paragraph tag? If someone's not using the existing callback I wouldn't expect them to provide any error handling. And I couldn't fault them for it, there's a whole laundry list of possibilities in the readme.

> I'm not sure that a replaceWith() method would make sense here. 

Err yeah you're right, I had it backwards in my head. I agree with deferring this though especially if anyone agrees with me on the above.

josiahsprague commented:
Added some tests @joanniclaborde :)

joanniclaborde commented on lib/wordpress.js:
```diff
@@ -0,0 +1,53 @@
+  params: !util.isNode?qs.parse(window.location.search.split('?').pop()):{},
+  getPostPreviewView: function (view) {
+    // Return false if we're not in preview mode
+    if (!this.params.is_preview) {
```
Unfortunately `this` will be wrong if you use the library like this:
```javascript
var getPostPreviewView = require('solidus-client/lib/wordpress').getPostPreviewView;
getPostPreviewView(view); // BLAMMO
```

You could call it like this instead: `if (!module.exports.params.is_preview) {`

Or, move the function to the util library (so it can be mocked during the tests):
```javascript
  getPostPreviewView: function (view) {
    // Return false if we're not in preview mode
    var params = util.params();
    if (!params.is_preview) {
```

joanniclaborde commented on test/tests/wordpress.js:
```diff
@@ -0,0 +1,132 @@
+        assert(typeof view.template == 'function');
+
+        // assert that
+          // all passed in keys are present on out
```
Forgot a test?

joanniclaborde commented:
Looking good, except this `params` thing. :+1:

josiahsprague commented:
Addressed feedback. Ready to do a release and deploy on https://github.com/SparkartGroupInc/jasonaldean.com/pull/77 if you guys wanna give me a green light. 🚦

pushred commented:
No comments here, next time please don't squash commits till right before you merge though, makes it difficult to review the progression of things without those diffs..

joanniclaborde commented on test/tests/wordpress.js:
```diff
@@ -0,0 +1,116 @@
+
+    describe('.getPostPreviewView', function() {
+      it('returns false when not in preview mode', function(done) {
+        util.params = {is_preview: false};
```
Make sure to reset the `params` object after each test, else it will break any other test that needs the object. See [example here](https://github.com/solidusjs/solidus-client/blob/master/test/tests/view.js#L218-L222).

joanniclaborde commented:
:+1: except that `params` comment.

josiahsprague commented:
How's that look @joanniclaborde ?

joanniclaborde commented:
:+1:
